File: 2641-cousins-in-binary-tree-ii/2641-cousins-in-binary-tree-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void UpdateTree(TreeNode* root, int level, vector<long long>& levelSum)
    {
        if(root==NULL)    return;
        int sum = 0;
        if(root->left)    sum += root->left->val;
        if(root->right)   sum += root->right->val;
        if(root->left)    root->left->val = levelSum[level+1] - sum;
        if(root->right)   root->right->val = levelSum[level+1] - sum;

        UpdateTree(root->left, level+1, levelSum);
        UpdateTree(root->right, level+1, levelSum);  
    }

    vector<long long> CalculateLevelSum(TreeNode* root)
    {
        vector<long long> levelSum;
        queue<TreeNode *> q; 
        q.push(root);

        while(!q.empty())
        {
            int n = q.size();
            int sum = 0;
            for(int i=0; i<n; i++)
            {
                TreeNode* curr = q.front();
                q.pop();
                sum += curr->val;
                if(curr->left)    q.push(curr->left);
                if(curr->right)   q.push(curr->right);
            }

            levelSum.push_back(sum);
        }

        return levelSum;
    }

    TreeNode* replaceValueInTree(TreeNode* root) {
        vector<long long> levelSum = CalculateLevelSum(root);
        int level = 0;
        UpdateTree(root, level, levelSum);
        root->val = 0;
        return root;
    }
};

File: 2641-cousins-in-binary-tree-ii/2641-cousins-in-binary-tree-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode replaceValueInTree(TreeNode root) {
        Deque<Node> queue = new ArrayDeque<>();
        int prevTotalSum=root.val;
        queue.add(new Node(root, root.val));

        while(!queue.isEmpty()){
            int size=queue.size();
            int totalSum=0;

            for(int i=0;i<size;i++){
                Node qNode = queue.poll();
                TreeNode node = qNode.node;
                int siblingSum = qNode.siblingSum;

                int sum=node.left!=null?node.left.val:0;
                sum+=node.right!=null?node.right.val:0;

                if(node.left!=null) queue.add(new Node(node.left, sum));
                if(node.right!=null) queue.add(new Node(node.right, sum));

                totalSum+=sum;
                node.val = prevTotalSum-siblingSum;
            }
            prevTotalSum=totalSum;
        }

        return root;
    }

    class Node{
        TreeNode node;
        int siblingSum;

        Node(TreeNode node, int siblingSum){
            this.node=node;
            this.siblingSum=siblingSum;
        }
    }
}